README
NAME
    Class::Accessor::FactoryTyped - Accessors whose values come from a
    factory

VERSION
    version 1.100880

SYNOPSIS
        package Person;
        use base 'Class::Accessor::FactoryTyped';

        __PACKAGE__->mk_factory_typed_accessors(
            'My::Factory',
            person_name    => 'name',
            person_address => 'address',
        );

DESCRIPTION
    This module generates accessors for your class in the same spirit as
    Class::Accessor does. While the latter deals with accessors for scalar
    values, this module provides accessor makers for arrays, hashes,
    integers, booleans, sets and more.

    As seen in the synopsis, you can chain calls to the accessor makers.
    Also, because this module inherits from Class::Accessor, you can put a
    call to one of its accessor makers at the end of the chain.

    The accessor generators also generate documentation ready to be used
    with Pod::Generated.

METHODS
  mk_factory_typed_accessors
        MyClass->mk_factory_typed_accessors(
            'My::Factory',
            foo => 'phooey',
            bar => [ qw(bar1 bar2 bar3) ],
            baz => {
                slot => 'foo',
                comp_mthds => [ qw(bar baz) ]
            },
            fob => [
                {
                    slot       => 'dog',
                    comp_mthds => 'bark',
                },
                {
                    slot       => 'cat',
                    comp_mthds => 'miaow',
                },
            ],
        );

    This behaves a lot like "Class::Accessor::Complex"'s
    "mk_object_accessors()", but the types of objects - that is, their class
    names - that the generated accessors can take aren't given statically,
    but are determined by asking a factory.

    The factory class name must be the first argument. The class indicated
    should be a subclass of Class::Factory::Enhanced.

    The following argument is an array which should contain pairs of class
    => sub-argument pairs. The sub-arguments are parsed like this:

    Hash Reference
        See "baz()" above. The hash should contain the following keys:

        "slot"
            The name of the instance attribute (slot).

        "comp_mthds"
            A string or array reference, naming the methods that will be
            forwarded directly to the object in the slot.

    Array Reference
        As for "String", for each member of the array. Also works if each
        member is a hash reference (see "fob()" above).

    String
        The name of the instance attribute (slot).

    For each slot "x", with forwarding methods "y()" and "z()", the
    following methods are created:

    "x" A get/set method, see "*" below.

    "y" Forwarded onto the object in slot "x", which is auto-created via
        "new()" if necessary. The "new()", if called, is called without
        arguments.

    "z" As for "y".

    So, using the example above, a method, "foo()", is created, which can
    get and set the value of those objects in slot "foo", which will
    generally contain an object of the type the factory, in this case
    "My::Factory", uses for the object type "baz". Two additional methods
    are created named "bar()" and "baz()" which result in a call to the
    "bar()" and "baz()" methods on the "Baz" object stored in slot "foo".

    Apart from the forwarding methods described above,
    "mk_object_accessors()" creates methods as described below, where "*"
    denotes the slot name.

    "*" If the accessor is supplied with an object of an appropriate type,
        will set set the slot to that value. Else, if the slot has no value,
        then an object is created by calling "new()" on the appropriate
        class, passing in any supplied arguments.

        The stored object is then returned.

    *_clear, "clear_*"
        Removes the object from the accessor.

  mk_factory_typed_array_accessors
    Like "mk_factory_typed_accessors()" except creates array accessors with
    all methods like those generated by "Class::Accessor::Complex"'s
    "mk_array_accessors()".

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Accessor-Factory
    Typed>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see
    <http://search.cpan.org/dist/Class-Accessor-FactoryTyped/>.

    The development version lives at
    <http://github.com/hanekomu/Class-Accessor-FactoryTyped/>. Instead of
    sending patches, please fork this project using the standard git and
    github infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Class-Accessor-FactoryTyped

1.100880  2010-03-29 19:58:20 Europe/Vienna
    - added t/03_array.t
    - use parent instead of use base
    - use Class::Accessor::Installer->document_accessor()

1.100830  2010-03-24 11:03:26 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.05  Fri 2010.02.05 22:10:05 CET (Marcel Gruenauer <marcel@cpan.org>)
     - check for defined() instead of ref()

0.04  Wed 2009.11.04 23:35:04 CET (Marcel Gruenauer <marcel@cpan.org>)
     - check whether what we want to check against UNIVERSAL::isa() is actually
       defined - in real-world scenarios, this saves most of the calls
     - Changed dist style and Changes back to standard. People didn't like it -
       the nail that sticks out gets hammered down.
     - Added standard test files; this will also help with CPANTS.

0.03  Thu, 29 May 2008 12:01:23 -0000 (Marcel Gruenauer <marcel@cpan.org>)
     - Converted Changes file to YAML style
     - .shipit: added Twitter step
     - Makefile.PL: added auto_install() and process_templates()
     - lib/*: converted to template
     - updated MANIFEST
     - .shipit: now uses ApplyYAMLChangeLogVersion
     - started writing generated documentation
     - updated t/perlcriticrc
     - factory typed attribute accessor now also accepts subclasses of the
       expected class
     - .shipit: fixed svk.tagpattern
     - tags: NEWFEATURE

0.02  Wed, 24 Oct 2007 11:41:27 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - now uses Class::Accessor::Installer

0.01  Sun, 21 Oct 2007 20:05:00 +0200 (Marcel Gruenauer <marcel@cpan.org>)
     - original version


t/lib/Person.pm
package Person;
our $VERSION = '1.100880';
use warnings;
use strict;
use parent 'Class::Accessor::FactoryTyped';
__PACKAGE__->mk_factory_typed_accessors(
    'MyFactory',
    person_name    => 'name',
    person_address => 'address',
);

__PACKAGE__->mk_factory_typed_array_accessors(
    'MyFactory',
    person_name => 'friends',
);
1;


t/lib/MyFactory.pm
package MyFactory;
our $VERSION = '1.100880';
use warnings;
use strict;
use parent 'Class::Factory::Enhanced';

# Default mappings
__PACKAGE__->register_factory_type(
    person_name    => 'Person::SimpleName',
    person_address => 'Person::SimpleAddress',
);
1;


inc/Test/Compile.pm
#line 1
package Test::Compile;

use warnings;
use strict;
use Test::Builder;
use Devel::CheckOS qw(os_is os_isnt);
use File::Spec;
use UNIVERSAL::require;


our $VERSION = '0.07';


my $Test = Test::Builder->new;


sub import {
    my $self = shift;
    my $caller = caller;

   for my $func (qw(
       pm_file_ok pl_file_ok all_pm_files all_pl_files all_pm_files_ok
       all_pl_files_ok
       )) {

        no strict 'refs';
        *{$caller."::".$func} = \&$func;
    }

    $Test->exported_to($caller);
    $Test->plan(@_);
}


sub pm_file_ok {
    my $file = shift;
    my $name = @_ ? shift : "Compile test for $file";

    if (!-f $file) {
        $Test->ok(0, $name);
        $Test->diag("$file does not exist");
        return;
    }

    my $module = $file;
    $module =~ s!^(blib/)?lib/!!;
    $module =~ s!/!::!g;
    $module =~ s/\.pm$//;

    my $ok = 1;
    $module->use;
    $ok = 0 if $@;

    my $diag = '';
    unless ($ok) {
        $diag = "couldn't use $module ($file): $@";
    }

    $Test->ok($ok, $name);
    $Test->diag($diag) unless $ok;
    $ok;
}


sub pl_file_ok {
    my $file = shift;
    my $name = @_ ? shift : "Compile test for $file";

    # Exclude VMS because $^X doesn't work
    # In general perl is a symlink to perlx.y.z
    # but VMS stores symlinks differently...
    unless (os_is('OSFeatures::POSIXShellRedirection') and os_isnt('VMS')) {
        $Test->skip('Test not compatible with your OS');
        return;
    }

    unless (-f $file) {
        $Test->ok(0, $name);
        $Test->diag("$file does not exist");
        return;
    }

    my $out = `$^X -cw $file 2>&1`;

    if ($?) {
        $Test->ok(0, 'Script does not compile');
        $Test->diag($out);
        return;
    } else {
        $Test->ok(1);
        return 1;
    }
}


sub all_pm_files_ok {
    my @files = @_ ? @_ : all_pm_files();

    $Test->plan(tests => scalar @files);

    my $ok = 1;
    for (@files) {
        pm_file_ok($_) or undef $ok;
    }
    $ok;
}


sub all_pl_files_ok {
    my @files = @_ ? @_ : all_pl_files();

    $Test->plan(tests => scalar @files);

    my $ok = 1;
    for (@files) {
        pl_file_ok($_) or undef $ok;
    }
    $ok;
}


sub all_pm_files {
    my @queue = @_ ? @_ : _pm_starting_points();
    my @pm;

    while (@queue) {
        my $file = shift @queue;
        if (-d $file) {
            local *DH;
            opendir DH, $file or next;
            my @newfiles = readdir DH;
            closedir DH;

            @newfiles = File::Spec->no_upwards(@newfiles);
            @newfiles = grep { $_ ne "CVS" && $_ ne ".svn" } @newfiles;

            for my $newfile (@newfiles) {
                my $filename = File::Spec->catfile($file, $newfile);
                if (-f $filename) {
                    push @queue, $filename;
                } else {
                    push @queue, File::Spec->catdir($file, $newfile);
                }
            }
        }
        if (-f $file) {
            push @pm, $file if $file =~ /\.pm$/;
        }
    }
    @pm;
}


sub all_pl_files {
    my @queue = @_ ? @_ : _pl_starting_points();
    my @pl;

    while (@queue) {
        my $file = shift @queue;
        if (-d $file) {
            local *DH;
            opendir DH, $file or next;
            my @newfiles = readdir DH;
            closedir DH;

            @newfiles = File::Spec->no_upwards(@newfiles);
            @newfiles = grep { $_ ne "CVS" && $_ ne ".svn" } @newfiles;

            for my $newfile (@newfiles) {
                my $filename = File::Spec->catfile($file, $newfile);
                if (-f $filename) {
                    push @queue, $filename;
                } else {
                    push @queue, File::Spec->catdir($file, $newfile);
                }
            }
        }
        if (-f $file) {
            # Only accept files with no extension or extension .pl
            push @pl, $file if $file =~ /(?:^[^.]+$|\.pl$)/;
        }
    }
    @pl;
}


sub _pm_starting_points {
    return 'blib' if -e 'blib';
    return 'lib';
}


sub _pl_starting_points {
    return 'script' if -e 'script';
    return 'bin' if -e 'bin';
}


1;


__END__

{% USE p = PodGenerated %}

#line 363



lib/Class/Accessor/FactoryTyped.pm
use 5.008;
use strict;
use warnings;

package Class::Accessor::FactoryTyped;
our $VERSION = '1.100880';

# ABSTRACT: Accessors whose values come from a factory
use Carp 'croak';
use Data::Miscellany 'set_push';
use UNIVERSAL::require;
use parent qw(
  Class::Accessor::Complex
  Class::Accessor::Installer
);
__PACKAGE__->mk_class_array_accessors(
    qw(factory_typed_accessors factory_typed_array_accessors));

sub mk_factory_typed_accessors {
    my ($self, $factory_class_name, @args) = @_;
    my $class = ref $self || $self;
    $factory_class_name->require or die $@;
    while (@args) {
        my $type = shift @args;
        my $list = shift @args or die "No slot names for $class";

        # Allow a list of hashrefs.
        my @list = (ref($list) eq 'ARRAY') ? @$list : ($list);
        for my $obj_def (@list) {
            my ($name, @composites);
            if (!ref $obj_def) {
                $name = $obj_def;
            } else {
                $name = $obj_def->{slot};
                my $composites = $obj_def->{comp_mthds};
                @composites =
                    ref($composites) eq 'ARRAY' ? @$composites
                  : defined $composites ? ($composites)
                  :                       ();
            }
            for my $meth (@composites) {
                $self->install_accessor(
                    name => $meth,
                    code => sub {
                        local $DB::sub = local *__ANON__ = "${class}::{$meth}"
                          if defined &DB::DB && !$Devel::DProf::VERSION;
                        my ($self, @args) = @_;
                        $self->$name()->$meth(@args);
                    },
                );
                $self->document_accessor(
                    name $meth,
                    purpose => <<EODOC,
Calls $meth() with the given arguments on the object stored in the $name slot.
If there is no such object, a new $type object is constructed - no arguments
are passed to the constructor - and stored in the $name slot before forwarding
$meth() onto it.
EODOC
                    examples => [ "\$obj->$meth(\@args);", "\$obj->$meth;", ],
                );
            }
            my $expected_class;

            # use a class list to the target package to keep track of which
            # framework_objects the class has, for introspection purposes
            $self->factory_typed_accessors_push($name);
            $self->install_accessor(
                name => $name,
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${name}"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self, @args) = @_;
                    unless ($expected_class) {
                        $expected_class =
                          $factory_class_name->get_registered_class($type);
                        die "no factory class name for type [$type]"
                          unless $expected_class;

                      # need to load the class to do UNIVERSAL::isa stuff on the
                      # class name
                        $expected_class->require or die $@;
                    }

                    # if (ref $args[0] eq $expected_class) {
                    if (defined($args[0])
                        && UNIVERSAL::isa($args[0], $expected_class)) {
                        return $self->{$name} = $args[0];
                    } elsif (@args || !defined $self->{$name}) {

                      # We accept a hashref of args as well and have to deref it
                      # first, since we're going to push args onto the @args
                      # array.
                        @args =
                          (scalar(@args == 1) && ref($args[0]) eq 'HASH')
                          ? %{ $args[0] }
                          : @args;

                      # Create an object if args are given, or autovivify one if
                      # no args are given and it doesn't exist yet.
                        return $self->{$name} =
                          $factory_class_name->make_object_for_type($type,
                            @args);
                    }

                # Still here? Hm, shouldn't happen, but return the value anyway.
                    $self->{$name};
                }
            );
            $self->install_accessor(
                name => [ "clear_${name}", "${name}_clear" ],
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${name}_clear"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    $_[0]->{$name} = undef;
                }
            );
            $self->install_accessor(
                name => [ "exists_${name}", "${name}_exists" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${name}_exists"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    exists $_[0]->{$name};
                }
            );
        }
    }
    $self;    # for chaining
}

sub mk_factory_typed_array_accessors {
    my ($self, $factory_class_name, @args) = @_;
    my $class = ref $self || $self;
    $factory_class_name->require or die $@;
    while (@args) {
        my $object_type_const = shift @args;
        my $list = shift @args or die "No slot names for $class";

        # Allow a list of hashrefs.
        my @list = (ref($list) eq 'ARRAY') ? @$list : ($list);
        for my $field (@list) {
            my $normalize = "${field}_normalize";
            $self->install_accessor(
                name => $normalize,
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${normalize}"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my $self = shift;
                    map {
                        ref $_
                          ? $_
                          : $factory_class_name->make_object_for_type(
                            $object_type_const, $_,);
                      }
                      map {
                        ref $_ eq 'ARRAY' ? @$_ : ($_)
                      } @_;
                }
            );

            # use a class list to the target package to keep track of which
            # framework_list_objects the class has, for introspection purposes
            $self->factory_typed_array_accessors_push($field);
            $self->install_accessor(
                name => $field,
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my $self = shift;
                    defined $self->{$field} or $self->{$field} = [];
                    @{ $self->{$field} } = $self->$normalize(@_) if @_;
                    wantarray ? @{ $self->{$field} } : $self->{$field};
                }
            );
            $self->install_accessor(
                name => [ "pop_${field}", "${field}_pop" ],
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}_pop"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self) = @_;
                    pop @{ $self->{$field} };
                }
            );
            $self->install_accessor(
                name => [ "set_push_${field}", "${field}_set_push" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_set_push"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self, @values) = @_;
                    set_push @{ $self->{$field} }, $self->$normalize(@values);
                }
            );
            $self->install_accessor(
                name => [ "push_${field}", "${field}_push" ],
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}_push"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self, @values) = @_;
                    push @{ $self->{$field} }, $self->$normalize(@values);
                }
            );
            $self->install_accessor(
                name => [ "shift_${field}", "${field}_shift" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_shift"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self) = @_;
                    shift @{ $self->{$field} };
                }
            );
            $self->install_accessor(
                name => [ "unshift_${field}", "${field}_unshift" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_unshift"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self, @values) = @_;
                    unshift @{ $self->{$field} }, $self->$normalize(@values);
                }
            );
            $self->install_accessor(
                name => [ "splice_${field}", "${field}_splice" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_splice"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self, $offset, $len, @list) = @_;
                    splice(@{ $self->{$field} }, $offset, $len, @list);
                }
            );
            $self->install_accessor(
                name => [ "clear_${field}", "${field}_clear" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_clear"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self) = @_;
                    @{ $self->{$field} } = ();
                }
            );
            $self->install_accessor(
                name => [ "count_${field}", "${field}_count" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_count"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self) = @_;
                    exists $self->{$field} ? scalar @{ $self->{$field} } : 0;
                }
            );
            $self->install_accessor(
                name => [ "index_${field}", "${field}_index" ],
                code => sub {
                    local $DB::sub = local *__ANON__ =
                      "${class}::${field}_index"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my $self = shift;
                    my (@indices) = @_;
                    my @Result;
                    push @Result, $self->{$field}->[$_] for @indices;
                    return $Result[0] if @_ == 1;
                    wantarray ? @Result : \@Result;
                }
            );
            $self->install_accessor(
                name => [ "set_${field}", "${field}_set" ],
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}_set"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my $self = shift;
                    my @args = @_;
                    croak "${field}_set expects an even number of fields\n"
                      if @args % 2;
                    while (my ($index, $value) = splice @args, 0, 2) {
                        $self->{$field}->[$index] = $self->$normalize($value);
                    }
                    return @_ / 2;
                }
            );
            $self->install_accessor(
                name => [ "ref_${field}", "${field}_ref" ],
                code => sub {
                    local $DB::sub = local *__ANON__ = "${class}::${field}_ref"
                      if defined &DB::DB && !$Devel::DProf::VERSION;
                    my ($self) = @_;
                    $self->{$field};
                }
            );
        }
    }
    $self;    # for chaining
}
1;


__END__
=pod

=head1 NAME

Class::Accessor::FactoryTyped - Accessors whose values come from a factory

=head1 VERSION

version 1.100880

=head1 SYNOPSIS

    package Person;
    use base 'Class::Accessor::FactoryTyped';

    __PACKAGE__->mk_factory_typed_accessors(
        'My::Factory',
        person_name    => 'name',
        person_address => 'address',
    );

=head1 DESCRIPTION

This module generates accessors for your class in the same spirit as
L<Class::Accessor> does. While the latter deals with accessors for scalar
values, this module provides accessor makers for arrays, hashes, integers,
booleans, sets and more.

As seen in the synopsis, you can chain calls to the accessor makers. Also,
because this module inherits from L<Class::Accessor>, you can put a call
to one of its accessor makers at the end of the chain.

The accessor generators also generate documentation ready to be used with
L<Pod::Generated>.

=head1 METHODS

=head2 mk_factory_typed_accessors

    MyClass->mk_factory_typed_accessors(
        'My::Factory',
        foo => 'phooey',
        bar => [ qw(bar1 bar2 bar3) ],
        baz => {
            slot => 'foo',
            comp_mthds => [ qw(bar baz) ]
        },
        fob => [
            {
                slot       => 'dog',
                comp_mthds => 'bark',
            },
            {
                slot       => 'cat',
                comp_mthds => 'miaow',
            },
        ],
    );

This behaves a lot like C<Class::Accessor::Complex>'s
C<mk_object_accessors()>, but the types of objects - that is, their class
names - that the generated accessors can take aren't given statically, but are
determined by asking a factory.

The factory class name must be the first argument. The class indicated should
be a subclass of L<Class::Factory::Enhanced>.

The following argument is an array which should contain pairs of class =>
sub-argument pairs. The sub-arguments are parsed like this:

=over 4

=item Hash Reference

See C<baz()> above. The hash should contain the following keys:

=over 4

=item C<slot>

The name of the instance attribute (slot).

=item C<comp_mthds>

A string or array reference, naming the methods that will be forwarded
directly to the object in the slot.

=back

=item Array Reference

As for C<String>, for each member of the array. Also works if each member is a
hash reference (see C<fob()> above).

=item String

The name of the instance attribute (slot).

=back

For each slot C<x>, with forwarding methods C<y()> and C<z()>, the following
methods are created:

=over 4

=item C<x>

A get/set method, see C<*> below.

=item C<y>

Forwarded onto the object in slot C<x>, which is auto-created via C<new()> if
necessary. The C<new()>, if called, is called without arguments.

=item C<z>

As for C<y>.

=back

So, using the example above, a method, C<foo()>, is created, which can get and
set the value of those objects in slot C<foo>, which will generally contain an
object of the type the factory, in this case C<My::Factory>, uses for the
object type C<baz>. Two additional methods are created named C<bar()> and
C<baz()> which result in a call to the C<bar()> and C<baz()> methods on the
C<Baz> object stored in slot C<foo>.

Apart from the forwarding methods described above, C<mk_object_accessors()>
creates methods as described below, where C<*> denotes the slot name.

=over 4

=item C<*>

If the accessor is supplied with an object of an appropriate type, will set
set the slot to that value. Else, if the slot has no value, then an object is
created by calling C<new()> on the appropriate class, passing in any supplied
arguments.

The stored object is then returned.

=item C<*_clear>, C<clear_*>

Removes the object from the accessor.

=back

=head2 mk_factory_typed_array_accessors

Like C<mk_factory_typed_accessors()> except creates array accessors with all
methods like those generated by C<Class::Accessor::Complex>'s
C<mk_array_accessors()>.

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Class-Accessor-FactoryTyped>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Class-Accessor-FactoryTyped/>.

The development version lives at
L<http://github.com/hanekomu/Class-Accessor-FactoryTyped/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Class::Accessor::FactoryTyped',
                       'DISTNAME' => 'Class-Accessor-FactoryTyped',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Accessors whose values come from a factory',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100880',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Scalar::Util' => '0',
                                        'English' => '0',
                                        'parent' => '0',
                                        'Data::Miscellany' => '0',
                                        'FindBin' => '0',
                                        'Class::Accessor::Installer' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Class::Factory::Enhanced' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'UNIVERSAL::require' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



